yarn/src/test/scala/org/apache/spark/deploy/yarn/YarnShuffleEncryptionSuite.scala

```diff
@@ -0,0 +1,226 @@
+  val serializer = new KryoSerializer(new SparkConf())
+  val hashPartitioner = new HashPartitioner(NUM_PARTITITONS)
+  val taskMetrics = new TaskMetrics()
+  val spillFilesCreated = new LinkedList[File]()
+
+  var partitionSizesInMergedFile: Array[Long] = null
+  val ugi = UserGroupInformation.createUserForTesting("testuser", Array[String] {
+    "testgroup"
+  })
+
+  @Mock(answer = RETURNS_SMART_NULLS) private var blockManager: BlockManager = _
+  @Mock(answer = RETURNS_SMART_NULLS) private var diskBlockManager: DiskBlockManager = _
+  @Mock(answer = RETURNS_SMART_NULLS) private var taskContext: TaskContext = _
+  @Mock(answer = RETURNS_SMART_NULLS) private var blockResolver: IndexShuffleBlockResolver = _
+  @Mock(
+    answer = RETURNS_SMART_NULLS) private var shuffleDep: ShuffleDependency[Object, Object,
+      Object] = _
+
+  override def beforeAll(): Unit = {
+    MockitoAnnotations.initMocks(this)
+    System.setProperty("SPARK_YARN_MODE", "true")
+    ugi.doAs(new PrivilegedExceptionAction[Unit]() {
+      override def run(): Unit = {
+        initialShuffleWriter()
+      }
+    })
+  }
+
+  override def afterAll(): Unit = {
+    Utils.deleteRecursively(tempDir);
+    val leakedMemory = taskMemoryManager.cleanUpAllAllocatedMemory();
+    if (leakedMemory != 0) {
+      fail("Test leaked " + leakedMemory + " bytes of managed memory");
+    }
+  }
+
+  test("Test Yarn shuffle encryption basic write") {
+    val doAsAction = new PrivilegedExceptionAction[Unit] {
+      override def run(): Unit = {
+        testYarnShuffleEncryptionWrite()
+      }
+    }
+    ugi.doAs(doAsAction)
+  }
+
+  private[this] def testYarnShuffleEncryptionWrite(): Unit = {
+    val dataToWrite = new ArrayList[Product2[Object, Object]]()
+    for (i <- 1 to NUM_PARTITITONS) {
+      val current = i - 1
+      dataToWrite.add(
+        new Tuple2[Object, Object](current.asInstanceOf[Object], current.asInstanceOf[Object]))
+    }
+    val writer = createWriter(true)
+    writer.write(dataToWrite.iterator())
+    writer.stop(true)
+
+    assertEquals(HashMultiset.create(dataToWrite), HashMultiset.create(readRecordsFromFile()))
+  }
+
+  @throws(classOf[IOException])
+  private[this] def createWriter(isShuffleEncrypted: Boolean): UnsafeShuffleWriter[Object,
+      Object] = {
+    conf.set("spark.shuffle.encryption.enabled", String.valueOf(isShuffleEncrypted))
+    new UnsafeShuffleWriter[Object, Object](
+      blockManager,
+      blockResolver,
+      taskMemoryManager,
+      new SerializedShuffleHandle[Object, Object](0, 1, shuffleDep),
+      0, // map id
+      taskContext,
+      conf
+    )
+  }
+
+  private[this] def readRecordsFromFile(): List[Tuple2[Object, Object]] = {
+    val recordsList = new ArrayList[Tuple2[Object, Object]]()
+    var startOffset = 0L
+    for (i <- 1 to NUM_PARTITITONS) {
+      val partitionSize = partitionSizesInMergedFile(i - 1)
+      if (partitionSize > 0) {
+        var in: InputStream = new FileInputStream(mergedOutputFile)
```

Comment:
In the updated patch, I use ```BlockStoreShuffleReader``` to read the encrypted files by mocking some methods in it.